src/main/resources/Confluence/Macros/ConfluenceGallery.xml

```diff
@@ -0,0 +1,723 @@
+        <displayFormType>select</displayFormType>
+        <displayType>yesno</displayType>
+        <name>supportsInlineMode</name>
+        <number>5</number>
+        <prettyName>Supports inline mode</prettyName>
+        <unmodifiable>0</unmodifiable>
+        <classType>com.xpn.xwiki.objects.classes.BooleanClass</classType>
+      </supportsInlineMode>
+      <visibility>
+        <cache>0</cache>
+        <disabled>0</disabled>
+        <displayType>select</displayType>
+        <freeText>forbidden</freeText>
+        <largeStorage>0</largeStorage>
+        <multiSelect>0</multiSelect>
+        <name>visibility</name>
+        <number>6</number>
+        <prettyName>Macro visibility</prettyName>
+        <relationalStorage>0</relationalStorage>
+        <separator>|</separator>
+        <separators>|</separators>
+        <size>1</size>
+        <unmodifiable>0</unmodifiable>
+        <values>Current User|Current Wiki|Global</values>
+        <classType>com.xpn.xwiki.objects.classes.StaticListClass</classType>
+      </visibility>
+    </class>
+    <property>
+      <async_cached>0</async_cached>
+    </property>
+    <property>
+      <async_context/>
+    </property>
+    <property>
+      <async_enabled>0</async_enabled>
+    </property>
+    <property>
+      <code>{{velocity}}
+## Get the parameters and put them in lists if necessary
+#set($title = $wikimacro.parameters.title)
+#set($includeString = $wikimacro.parameters.include)
+#set($excludeString = $wikimacro.parameters.exclude)
+#set($pagesString = $wikimacro.parameters.page)
+#set($include = $stringtool.split($includeString, ","))
+#set($exclude = $stringtool.split($excludeString, ","))
+#set($pages = $stringtool.split($pagesString, ","))
+## Display the gallery title if it is set (newlines are needed for the following macro calls to be standalone)
+#if(!$stringtool.isEmpty($title))
+    ==$title==
+    
+    
+#end
+## add the current page to the page list if no pages were given
+#if($stringtool.isEmpty($pagesString))
+    #set ($pages = [$doc.getTitle()])
+#end
+## Collect the attachments to display
+#set ($attachments = [])
+#foreach($page in $pages)
+    ## Resolve page and put attachments in temporary list
+    #set ($tempAttachments = [])
+    #set($query = $services.query.xwql("where doc.title = :name").bindValue('name', ''+${page}))
+    #set($pageStrings = $query.execute())
+    #if ($pageStrings.size() &gt; 0)
+        #set($pageDocument = $xwiki.getDocument($pageStrings[0]))
+        #set($discard = $tempAttachments.addAll($pageDocument.getAttachmentList()))
+        ## iterate over remaining attachments and add full reference to attachment list
+        #foreach($attachment in $tempAttachments)
+            #if($attachment.isImage())
+                ## If includes or excludes are set, remove elements from temporary attachment list
+                #if(!$stringtool.isEmpty($includeString))
+                    #if ($include.contains($attachment.filename))
```

Comment:
Why not grouping these 2 if statements in `#if(!$stringtool.isEmpty($includeString) && $include.contains($attachment.filename)))`
Even better, you could check if the `$includeString` is empty when assigning `$include`, then check only if the $include contains the filename:
```
#set ($includeString = $wikimacro.parameters.include)
...
#if ("$!includeString" != '')
  #set ($include = $includeString.split(','))
#end
...
#if ($include.contains($attachment.filename))
  #set ($discard = $attachments.add($pageDocument + '@' + $attachment.filename))
#end
```

Reply:
Will velocity simply skip the `#if ($include.contains($attachment.filename))` clause when `$include` is not defined? Based on Java I would expect NullPointer issues, but my Velocity knowledge is very limited.

Reply:
Yes, it will silently fail, meaning that if `$include` is not defined, the evaluation result in `false`.

Reply:
Okay, thanks. On rereading the code I remember that I not group the ifs on purpose, as I only wanted to handle excludes if no includes are specified (includes taking precedence over excludes). If includes are specified everything else is excluded anyway, so no need to extra handle excludes (unless someone wants to exclude an included image, which I see no sense in). Excludes are then only handled if there are no includes. If I group the statements there may be the case that the include string is not empty, but the current attachment is not in it. In that case the macro would then proceed to (unnecessarily) check if the image should be excluded. The same would happen for the second approach you posted. It does not make that much of a difference, I just wanted to avoid unnecessary calls. If you want I can rewrite it anyway.